Storage/Storage/Tools/TypedPredicates.swift

```diff
@@ -0,0 +1,80 @@
+
+/// Overloads for compound operators.
+/// Each operator takes `TypedPredicate` instances and returns a `CompoundPredicate` that evaluate the predicates using their respective logical rule.
+///
+
+/// Overloads the `AND` operator between two predicates, returns a `CompoundPredicate` that evaluates the two predicates with the `&&` rule.
+///
+public func && (p1: TypedPredicate, p2: TypedPredicate) -> CompoundPredicate {
+    CompoundPredicate(type: .and, subpredicates: [p1, p2])
+}
+
+/// Overloads the `OR` operator between two predicates, returns a `CompoundPredicate` that evaluates the two predicates with the `!!` rule.
+///
+public func || (p1: TypedPredicate, p2: TypedPredicate) -> CompoundPredicate {
+    CompoundPredicate(type: .or, subpredicates: [p1, p2])
+}
+
+// MARK: - Comparison Operators Overloads
+
+/// Overloads for comparison  operators.  Each operator takes a `KeyPath` and a `Value`
+/// That will be mapped to `NSExpressions` instances  for then constructing a `ComparisonPredicate` with the correct comparison rule.
+/// These overloads make use of `Generics` because need to know at compile time that the `KeyPath.ResultingType` has the same type as the `Value`.
+/// As well as making sure that those values are `Equatable`.
+
+/// Overloads the `equal` operator between a `KeyPath` and a `Value`. Returns a `ComparisonPredicate` that evaluates the parameters using the `==` rule.
+///
+public func == <RootType, ResultingType: Equatable>(keyPath: KeyPath<RootType, ResultingType>, value: ResultingType) -> ComparisonPredicate {
+    ComparisonPredicate(keyPath, .equalTo, value)
+}
+
+/// Overloads the `not equal` operator between a `KeyPath` and a `Value`. Returns a `ComparisonPredicate` that evaluates the parameters using the `!=` rule.
+///
+public func != <RootType, ResultingType: Equatable>(keyPath: KeyPath<RootType, ResultingType>, value: ResultingType) -> ComparisonPredicate {
+    ComparisonPredicate(keyPath, .notEqualTo, value)
+}
+
+/// Overloads the  `IN` operator between a `KeyPath` and a `[Value]`. Returns a `ComparisonPredicate` that evaluates the parameters using the `===` rule.
+///
+public func === <RootType, ResultingType: Equatable>(keyPath: KeyPath<RootType, ResultingType>, value: Array<ResultingType>) -> ComparisonPredicate {
+    ComparisonPredicate(keyPath, .in, value)
+}
+
+// MARK: - Convenience Initializers
+
+internal extension ComparisonPredicate {
+    /// Returns a `ComparisonPredicate` by converting the parameters into `NSExpression` instances using the provided `Operator` as a modifier.
+    ///
+    convenience init<RootType, ResultingType>(_ keyPath: KeyPath<RootType, ResultingType>, _ operator: NSComparisonPredicate.Operator, _ value: Any?) {
+        let keyPathExpression = NSExpression(forKeyPath: keyPath)
+        let valueExpression = NSExpression(forConstantValue: value)
+        let options = ResultingType.self == String.self ? NSComparisonPredicate.Options.caseInsensitive : []
```

Comment:
While reviewing https://github.com/woocommerce/woocommerce-ios/pull/3487, I noticed that some string comparisons aren't case insensitive (like `loadTaxClass(slug:)`) - is it possible to override `NSComparisonPredicate.Options` in some way?

Reply:
Wow, great catch I failed to notice that case! To solve it I have added a new operator `=~` that performs a `==` comparison with the `case insensitive` option. https://github.com/woocommerce/woocommerce-ios/pull/3485/commits/de1a07cfc6033c807ec233197bf89c7a23fb32e4